### PR TITLE
Improve ChangeTracking performance

### LINKED ISSUE #11194: Investigate perf regressions from 2.0 related to change tracking
Specifically:
* Add/Attach/Update/Remove including Range, Children, and Parent variations
* QueryChildren

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #11194

### CODE REVIEW

File: src/EFCore/ChangeTracking/Internal/EntityGraphAttacher.cs [hunk 1]
```diff
@@ -31,10 +37,10 @@ public virtual void AttachGraph(InternalEntityEntry rootEntry, EntityState entit
             => _graphIterator.TraverseGraph(
                 new EntityEntryGraphNode(rootEntry, null, null)
                 {
-                    NodeState = entityState
+                    NodeState = _boxedEntityStates[(int)entityState]
```
Comment (smitpatel) [inline src/EFCore/ChangeTracking/Internal/EntityGraphAttacher.cs]: Fancy

File: src/EFCore/ChangeTracking/Internal/ArrayPropertyValues.cs [hunk 2]
```diff
@@ -54,7 +54,7 @@ public override void SetValues(object obj)
                 {
                     if (!Properties[i].IsShadowProperty)
                     {
-                        SetValue(i, Properties[i].GetGetter().GetClrValue(obj));
+                        SetValue(i, ((Property)Properties[i]).Getter.GetClrValue(obj));
```
Comment (smitpatel) [inline src/EFCore/ChangeTracking/Internal/ArrayPropertyValues.cs]: Is it safe to cast?

File: src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs [hunk 3]
```diff
@@ -582,7 +605,7 @@ protected virtual void WritePropertyValue([NotNull] IPropertyBase propertyBase,
         {
             Debug.Assert(!propertyBase.IsShadowProperty);
 
-            propertyBase.GetSetter().SetClrValue(Entity, value);
+            ((PropertyBase)propertyBase).GetSetter().SetClrValue(Entity, value);
```
Comment (smitpatel) [inline src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs]: Can you just Call member Setter

File: src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs [hunk 4]
```diff
@@ -626,7 +649,7 @@ public virtual void RemoveFromCollection([NotNull] INavigation navigation, [NotN
         {
             Debug.Assert(!navigation.IsShadowProperty);
 
-            navigation.GetCollectionAccessor().Remove(Entity, value.Entity);
+            ((Navigation)navigation).GetCollectionAccessor().Remove(Entity, value.Entity);
```
Comment (smitpatel) [inline src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs]: Member access?

File: src/EFCore/ChangeTracking/Internal/ArrayPropertyValues.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/ChangeTracking/Internal/ArrayPropertyValues.cs]: All the added casts are safe for now as we don't support other model implementations

**Pull request merged**